scripts/content_type_stats.py

```diff
@@ -0,0 +1,134 @@
+        list_of_dirs = list_of_dirs[: upto_i + 1]
+    if from_dir:
+        try:
+            from_i = [dir.name for dir in list_of_dirs].index(from_dir)
+        except ValueError:
+            raise ValueError('The from-dir is not a valid batch')
+        list_of_dirs = list_of_dirs[from_i:]
+    return list_of_dirs
+
+
+def process_batch(input_dir, output_dir):
+    """
+    Processes a batch, collecting mime-type data from its files.
+    Writes the results into a single tsv.gz file.
+    """
+    stat_file = (output_dir / input_dir.name).with_suffix('.tsv.gz')
+    matcher_ct = re.compile(r'Content-Type: ([\w/]+)', re.I)
+    matcher_wi = re.compile(r'WARC-IDentified-Payload-Type: ([\w/]+)', re.I)
+    matcher_cd = re.compile(r'(Content-Disposition: )([^\n^"]*)"([^\n^"]*)"', re.I)
+    with openall(stat_file, "wt", encoding="utf-8") as f:
+        for file in input_dir.iterdir():
+            file_name = input_dir.name + '/' + file.name
+            for doc in parse_file(file):
+                # Get the type from the metadata of the doc tag:
+                if doc.attrs['mime-type']:
+                    type_from_doctag = doc.attrs['mime-type']
+                else:
+                    type_from_doctag = '-'
+                # Get the type from the warc request header:
+                match_warcid = matcher_wi.search(doc.meta['request'])
+                if match_warcid:
+                    type_from_warc_id = match_warcid.groups(1)[0]
+                else:
+                    type_from_warc_id = '-'
+                # Get the type from the response header:
+                match_resp = matcher_ct.search(doc.meta['response'])
+                if match_resp:
+                    type_from_response = match_resp.groups(1)[0]
+                else:
+                    type_from_response = '-'
+                # Get the info about the attached file, if any:
+                match_cd = matcher_cd.search(doc.meta['response'])
+                if match_cd:
+                    attached_file = match_cd.group(3)
+                    if '.' in attached_file:
+                        attachment_ext = attached_file.split('.')[-1]
+                    else:
+                        attachment_ext = 'no_extension'
+                else:
+                    attachment_ext = '-'
+                # Process the results for this doc:
+                line = file_name + '\t'
+                line += type_from_doctag + '\t'
+                line += type_from_warc_id + '\t'
+                line += type_from_response + '\t'
+                line += attachment_ext + '\t'
+                print(line, file=f)
+
+
+def main():
+    args = parse_arguments()
+
+    logging.basicConfig(
+        level=getattr(logging, args.log_level.upper()),
+        format='%(asctime)s - %(process)s - %(levelname)s - %(message)s'
+    )
+
+    task_dirs = fetch_directories(args.input_dir, args.from_dir, args.upto_dir)

```

Comment:
Let's keep it simpler. Drop the from and to dir arguments, and parallelize the content type extraction instead. As a rule, there won't be too many documents in a single file, so we won't run into memory problems. We can easily implement the outer for loop (which is in the pool now) in the shell.

Reply:
The outer loop goes through the batches.
The inner loop goes through the .txt.gz files.

We want to output one statistics file per batch (not one per .txt.gz file!).

So if we parallelize the inner loop it means that multiple processes write into the same file, appending lines one by one. 
I did a quick googling for this and they say that on unix it can work without using locking... are we okay with that? 

It also means that the lines in the stats files will be jumbled...